UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs

```diff
@@ -19,468 +12,138 @@ namespace MLAgents
+            _inferenceOutputs = _metaDataLoader.GetOutputTensors();
+            _tensorGeneratorInvoker = new TensorGeneratorInvoker(brainParameters, seed);
+            _outputTensorApplierInvoker = new TensorApplierInvoker(brainParameters, seed);
+        }
+        
+        /// <summary>
+        /// Return a list of failed checks corresponding to the failed compatibility checks
+        /// between the Model and the BrainParameters. Note : This does not reload the model.
+        /// If changes have been made to the BrainParameters or the Model, the model must be
+        /// reloaded using GiveModel before trying to get the compatibility checks.
+        /// </summary>
+        /// <returns> The list of the failed compatibility checks between the Model and the
+        /// Brain Parameters</returns>
+        public IEnumerable<string> GetModelFailedChecks()
+        {
+            return (_metaDataLoader != null) ? _metaDataLoader.GetChecks() : new List<string>();
         }
 
-
-        /// Uses the stored information to run the tensorflow graph and generate 
-        /// the actions.
+        /// <inheritdoc />
         protected override void DecideAction()
         {
-#if ENABLE_TENSORFLOW
             base.DecideAction();
-            
-            if (isControlled)
+            if (_isControlled)
             {
                 agentInfos.Clear();
                 return;
             }
-
-            int currentBatchSize = agentInfos.Count();
-            List<Agent> agentList = agentInfos.Keys.ToList();
+            var currentBatchSize = agentInfos.Count();
             if (currentBatchSize == 0)
             {
                 return;
             }
             
-
-            // Create the state tensor
-            if (hasState)
+            // Prepare the input tensors to be feed into the engine
+            foreach (var tensor in _inferenceInputs)
             {
-                int stateLength = 1;
-                stateLength = brainParameters.vectorObservationSize;
-                inputState =
-                    new float[currentBatchSize, stateLength * brainParameters.numStackedVectorObservations];
-
-                var i = 0;
-                foreach (Agent agent in agentList)
+                if (!_tensorGeneratorInvoker.ContainsKey(tensor.Name))
                 {
-                    List<float> stateList = agentInfos[agent].stackedVectorObservation;
-                    for (int j =
-                            0;
-                        j < stateLength * brainParameters.numStackedVectorObservations;
-                        j++)
-                    {
-                        inputState[i, j] = stateList[j];
-                    }
-
-                    i++;
+                    throw new UnityAgentsException(
+                        "Unknow tensor expected as input : "+tensor.Name);
                 }
+                _tensorGeneratorInvoker[tensor.Name].Execute(tensor, currentBatchSize, agentInfos);
```

Comment:
Seeing how it's used here, there's a way to simplify even more. Having a class named "Invoker" is a hint that something is wrong (or else, why not invoke directly). TensorGenerator interface can be a private interface "Generator" inside the Invoker, and then the Invoker can be the "TensorGenerator". 

Another thing, if you move that loop into the TensorGeneratorInvoker, then the Invoker doesn't need to re-implement ContainsKey and [] and the next loop for the outputs can disappear too and we'll have even less copy paster. 

Also "Generate" is a better name than Execute here.